File: training-data/ruby/bipartite_check.rb
require 'set'

# A graph is bipartite iff it can be 2-colored so that no edge joins
# two nodes of the same color. BFS assigns alternating colors and
# fails as soon as an edge would force two neighbors to share one.
def bipartite?(graph)
  colors = {}

  graph.each_key do |start|
    next if colors.key?(start)

    colors[start] = 0
    queue = [start]

    until queue.empty?
      node = queue.shift
      graph[node].each do |neighbor|
        if colors.key?(neighbor)
          return false if colors[neighbor] == colors[node]
        else
          colors[neighbor] = 1 - colors[node]
          queue << neighbor
        end
      end
    end
  end

  true
end

even_cycle = { 0 => [1, 3], 1 => [0, 2], 2 => [1, 3], 3 => [2, 0] }
odd_cycle = { 0 => [1, 2], 1 => [0, 2], 2 => [0, 1] }

puts bipartite?(even_cycle) # true
puts bipartite?(odd_cycle)  # false

File: training-data/perl/bipartite_check.pl
use strict;
use warnings;

sub is_bipartite {
    my ($graph) = @_;
    my %color;
    for my $start (keys %$graph) {
        next if exists $color{$start};
        $color{$start} = 0;
        my @queue = ($start);
        while (@queue) {
            my $u = shift @queue;
            for my $v (@{$graph->{$u}}) {
                if (!exists $color{$v}) {
                    $color{$v} = 1 - $color{$u};
                    push @queue, $v;
                } elsif ($color{$v} == $color{$u}) {
                    return 0;
                }
            }
        }
    }
    return 1;
}

my $even_cycle = { a => ['b', 'd'], b => ['a', 'c'], c => ['b', 'd'], d => ['a', 'c'] };
my $odd_cycle  = { a => ['b', 'c'], b => ['a', 'c'], c => ['a', 'b'] };

print is_bipartite($even_cycle) ? "true" : "false", "\n";
print is_bipartite($odd_cycle) ? "true" : "false", "\n";
